js/ui/paygUnlockDialog.js

```diff
@@ -0,0 +1,395 @@
+            return Clutter.EVENT_PROPAGATE;
+
+        // Don't allow inserting more digits than required.
+        if (this._code.length >= CODE_ENTRY_REQUIRED_LENGTH)
+            return Clutter.EVENT_STOP;
+
+        // Allow digits only
+        if (GLib.unichar_isdigit(character))
+            this.clutter_text.insert_unichar(character);
+
+        return Clutter.EVENT_STOP;
+    },
+
+    _onTextChanged: function(textActor) {
+        this._code = textActor.text;
+        this.emit('code-changed', this._code);
+    },
+
+    _onButtonPressEvent: function() {
+        this.grab_key_focus();
+        return false;
+    },
+
+    addCharacter: function(character) {
+        this.clutter_text.insert_unichar(character);
+    },
+
+    setEnabled: function(value) {
+        this.reactive = value;
+        this.clutter_text.editable = value;
+    },
+
+    reset: function() {
+        this.text = '';
+    },
+
+    get code() {
+        return this._code;
+    },
+
+    get length() {
+        return this._code.length;
+    }
+});
+
+var PaygUnlockDialog = new Lang.Class({
+    Name: 'PaygUnlockDialog',
+
+    _init: function(parentActor) {
+        this._parentActor = parentActor;
+
+        this._entry = null;
+        this._errorMessage = null;
+        this._cancelButton = null;
+        this._nextButton = null;
+        this._spinner = null;
+        this._cancelled = false;
+
+        this._verificationStatus = UnlockStatus.NOT_VERIFYING;
+
+        // Clear the clipboard to make sure nothing can be copied into the entry.
+        St.Clipboard.get_default().set_text(St.ClipboardType.CLIPBOARD, '');
+        St.Clipboard.get_default().set_text(St.ClipboardType.PRIMARY, '');
+
+        this.actor = new St.Widget({ accessible_role: Atk.Role.WINDOW,
+                                     style_class: 'unlock-dialog-payg',
+                                     layout_manager: new Clutter.BoxLayout(),
+                                     visible: false });
+        this.actor.add_constraint(new Monitor.MonitorConstraint({ primary: true }));
+
+        this._parentActor.add_child(this.actor);
+
+        let promptBox = new St.BoxLayout({ vertical: true,
+                                           x_align: Clutter.ActorAlign.CENTER,
+                                           y_align: Clutter.ActorAlign.CENTER,
+                                           x_expand: true,
+                                           y_expand: true,
+                                           style_class: 'unlock-dialog-payg-layout'});
+        promptBox.connect('key-press-event', (actor, event) => {
+            if (event.get_key_symbol() == Clutter.KEY_Escape)
+                this._onCancelled();
+
+            return Clutter.EVENT_PROPAGATE;
+        });
+        this.actor.add_child(promptBox);
+
+        let titleLabel = new St.Label({ style_class: 'unlock-dialog-payg-title',
+                                        text: _("Your subscription has expired!"),
```

Comment:
Nitpick: Unless this was mandated by the designers, I’d drop the `!`. It seems a bit excited and unprofessional for something which is essentially about to ask the user for money.

Reply:
This was basically a placeholder, meant to be replaced with whatever we agreed upon, and the `!` was a reminder of that temporary nature. But I will remove it, in case it becomes the final thing.

Reply:
Ah, I see. I would have left a comment like `/* FIXME: Put final UI strings in place */` or summat.